luacompiler/src/lib/irgen/register_map.rs

```diff
@@ -1,43 +1,80 @@
-use std::{cell::RefCell, collections::HashMap};
+use std::{collections::HashMap, vec::Vec};
 
-/// Represents a structure that is used to keep track of the mapping
-/// between Lua variables and register ids.
-pub struct RegisterMap {
-    reg_count: RefCell<u8>,
-    reg_map: RefCell<HashMap<String, u8>>,
+/// Represents a tuple which is used to specify the lifetime of a register.
+/// For example if a register is first used by the 4th instruction of the bytecode, and
+/// used last by the 7th instruction, the register's lifetime would be (4, 8).
+pub struct Lifetime(usize, usize);
+
+impl Lifetime {
+    pub fn new(sp: usize) -> Lifetime {
+        Lifetime(sp, sp + 1)
+    }
+
+    /// Get the start point of the register.
+    pub fn start_point(&self) -> usize {
+        self.0
+    }
+
+    /// Get the end point of the register.
+    pub fn end_point(&self) -> usize {
+        self.1
+    }
+
+    fn set_end_point(&mut self, ep: usize) {
+        self.1 = ep
+    }
+}
+
+/// Represents a structure that is used to map Lua variables to registers, and to keep
+/// track of their lifetimes.
+pub struct RegisterMap<'a> {
+    current_instr: usize,
```

Comment:
Why is `current_instr` part of the register map? Isn't it the same as `lifetimes.len()`?

Reply:
Good spot! You are right, it is!